apps/service_providers/llm_service/adapters.py

```diff
@@ -119,17 +93,176 @@ def save_message_to_history(self, message: str, type_: ChatMessageType, experime
+            llm_service=experiment.get_llm_service(),
+            temperature=experiment.temperature,
+            prompt_text=experiment.prompt_text,
+            max_token_limit=experiment.max_token_limit,
+            experiment_version_number=experiment.version_number,
+            experiment_is_a_version=experiment.is_a_version,
+            tools=get_tools(session, experiment=experiment),
+            input_formatter=experiment.input_formatter,
+            source_material_id=experiment.source_material_id,
+            trace_service=trace_service,
+        )
+
+    @staticmethod
+    def from_pipeline(experiment: Experiment, session: ExperimentSession) -> Self:
+        """TODO"""
+
+    def get_chat_model(self):
+        return self.get_llm_service().get_chat_model(self.provider_model_name, self.temperature)
+
+    def get_template_context(self, variables: list[str]):
+        return self.template_context.get_context(variables)
+
+    def get_prompt(self):
+        return self.prompt_text
+
+    def get_chat_history(self, input_messages: list):
+        return compress_chat_history(
+            self.session.chat,
+            llm=self.get_chat_model(),
+            max_token_limit=self.max_token_limit,
+            input_messages=input_messages,
+        )
+
     def check_cancellation(self):
-        self.chat.refresh_from_db(fields=["metadata"])
+        self.session.chat.refresh_from_db(fields=["metadata"])
         # temporary mechanism to cancel the chat
         # TODO: change this to something specific to the current chat message
-        return self.chat.metadata.get("cancelled", False)
+        return self.session.chat.metadata.get("cancelled", False)
+
+
+class AssistantAdapter(BaseAdapter):
+    def __init__(
+        self,
+        session: ExperimentSession,
+        assistant: OpenAiAssistant,
+        citations_enabled: bool,
+        experiment_version_number: int,
+        experiment_is_a_version: bool,
+        input_formatter: str | None = None,
+        trace_service=None,
+        save_message_metadata_only: bool = False,
+    ):
+        self.session = session
+        self.team = session.team
+        self.assistant = assistant
+        self.llm_service = assistant.llm_provider.get_llm_service()
+        self.provider_model_name = assistant.llm_provider_model.name
+        self.citations_enabled = citations_enabled
+        self.experiment_version_number = experiment_version_number
+        self.experiment_is_a_version = experiment_is_a_version
+        self.tools = get_assistant_tools(assistant, experiment_session=session)
+        self.input_formatter = input_formatter
+        self.trace_service = trace_service
+        self.save_message_metadata_only = save_message_metadata_only
+        self.input_message_metadata = {}
+        self.output_message_metadata = {}
+        self.template_context = PromptTemplateContext(session, source_material_id=None)
```

Comment:
nit: (This is purely a stylistic thing but it does help with readability and cognitive overhead.)

Values coming from the method args should go above values that are computed. Also, where possible keep the ordering consistent between the args and the property assignment:

```python
def __init__(self, a, b, c):
    self.a = a
    self.b = b
    self.c = c
    self.service = a.get_service()
```
```